Fix login flow, headless mode and remove unnecessary extra options from code

robertoalvarezalonso opened the pull request:
ServiceNow recently updated their login page, which caused the existing selectors in the script to break. This PR updates the selectors for the logo, username field, next button, password field, and submit button to reflect the new DOM structure.

robertoalvarezalonso committed:
updated selectors to match new login page

robertoalvarezalonso committed:
Fixed headed mode

0x111 commented:
Thanks for the changes, I will review it in the next few days and let you know, I need to test it briefly.

0x111 commented:
Sorry for the delay, I am working on this now.

robertoalvarezalonso commented:
No worries, thanks for letting me know!
Take your time — if you spot anything that needs adjustment, just let me know and I’ll be happy to update it.

0x111 commented:
It is unbelievable, but my instance does not go to sleep and it has been at least a week. I will try on a different account. Did you have any experience like this before?

I see it the first time.

robertoalvarezalonso commented:
No, I haven't. In my case, it’s actually the opposite!

robertoalvarezalonso commented:
Just following up to see if you had the time to look at this PR. I've been using this new version for a while now without any problems.

0x111 commented on servicenow-instance-wakeup.go:
```diff
@@ -49,11 +50,11 @@ func main() {
 	}
 
-	opts := append(chromedp.DefaultExecAllocatorOptions[:],
+	opts := []chromedp.ExecAllocatorOption{
```
Why did you needed to change this please? (Just curious)

I tried changing it back now and for me it seems to work with the old one too, headless mode. Did you have some problems with it? Thanks

robertoalvarezalonso replied:
To be honest, I don’t quite remember why I made that change.
I was probably experimenting at the time and needed more explicit control over the options, rather than relying on the defaults.

Happy to stick with the previous version if you prefer it. I don’t see any downside.

Thanks for checking that!

0x111 replied:
I would keep the previous version if no problems with it please. I will then release new docker versions and app versions. Thank you

robertoalvarezalonso replied:
Sorry — I just remembered why I made that change. Non-headless mode wasn’t working with the previous setup, so I updated the code to fix that.

Thanks again for checking!

robertoalvarezalonso replied:
With this new commit, we have the headless and non-headless modes working, while maintaining the DefaultExecAllocatorOptions of the previous version.

0x111 replied:
I see, did you test it with the Dockerfile by building the Docker image? Because I just tried it with the old options, with your changes made in the code for the elements and it did run fine with the chromedp/headless-shell image that is used in the [Dockerfile](https://github.com/0x111/servicenow-instance-wakeup/blob/master/Dockerfile).

Feel free to try it:
1. Clone the repo
2. Run this command to build the docker image
```shell
docker build -t sniw .
```
3. Run the built docker image
```shell
docker run --rm -e USERNAME="DEVELOPER_USER" -e PASSWORD="DEVELOPER_PASSWORD." --name sniwt sniw
````

And you should see the output going through. Let me know if it works this way for you with the old options and your changes. Sorry for the "nitpicking", I just want to avoid changing something that works without any reason, this way we avoid future problems.

robertoalvarezalonso replied:
Thanks for the detailed steps!

I tested it using the Dockerfile and it works fine with the old and new flags in headless mode, as expected.

Just to clarify though: the issue isn’t with headless — it’s actually the opposite. When running in **non-headless** mode, the old options don’t behave as expected.

The goal of [this commit](https://github.com/0x111/servicenow-instance-wakeup/pull/48/commits/aaa7b1e2bd391ec24f7c1222e59526cbeeb99d2f) was precisely to keep the flags that were already in place, while making sure things also work when not running headless. This is especially helpful for debugging, even if the final use case remains headless.

As you can see here, the _DefaultExecAllocatorOptions_ already contains the _Headless_ option.
![image](https://github.com/user-attachments/assets/392fbacc-654c-434d-b24d-13bdf4c72dbd)

That’s why it wouldn’t work properly in non-headless mode without explicitly overriding it.
![image](https://github.com/user-attachments/assets/ba308700-a657-45d6-85bb-f4f5cdc6b646)

0x111 replied:
Thank you for the explanation, makes sense.

0x111 commented:
Sorry for the big delay, I had to make a new account and request an instance there, was in the "waiting queue" for a while. 
The instance on the other account bugged out somehow and was NEVER asleep...

Except of the one comment, LGTM

robertoalvarezalonso commented:
No worries at all. Thanks for the update!

And great to hear it looks good apart from that one comment. Let me know if anything else pops up!

robertoalvarezalonso committed:
fix: improve Chrome initialization configuration for (non)headless mode

Removed these two because they are already part of DefaultExecAllocatorOptions:
- chromedp.NoFirstRun
- chromedp.NoDefaultBrowserCheck

0x111 commented:
Thank you for your contribution, appreciate it a lot!

I will release a new version today or early tomorrow.

robertoalvarezalonso commented:
Thanks! This project has been super useful for me, so happy to give back!
